### ISSUE TITLE
Confusing error message in the case of callables

### ISSUE THREAD

**Issue by alex** [astral-sh/ty#3697]
### Summary

When there's a type error involving callables, the error message can be confusing -- I'm not sure exactly how general this problem is, but concrete reproducer:

```python
from collections.abc import Callable

def f() -> list[str]:
    return []

def takes(names: dict[str, Callable[[], list[str]]]) -> None:
    pass

d = {"f": f}
takes(d)
```

The error message here is:

```
error[invalid-argument-type]: Argument to function `takes` is incorrect
  --> t.py:10:7
   |
10 | takes(d)
   |       ^ Expected `dict[str, () -> list[str]]`, found `dict[str, () -> list[str]]`
   |
info: Function defined here
 --> t.py:6:5
  |
6 | def takes(names: dict[str, Callable[[], list[str]]]) -> None:
  |     ^^^^^ ----------------------------------------- Parameter declared here
  |
info: `dict` is invariant in its second type parameter
info: Consider using the supertype `collections.abc.Mapping`, which is covariant in its value type
info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics
```

Specifically the _expected_ line is incredibly confusing because the expected and found are the same.

### Version

0.0.45

**Comment by carljm** [comment]
Thanks for the report!

The cause is that we have an internal "function-like" flag to distinguish callable types which should have bound-method descriptor behavior vs callable types which should not. Currently we consider function-like callables to be a subtype of the regular callable with equivalent signature, thus they are not equivalent types, and thus the invariance error here. 

I assume you already discovered that an explicit annotation on `d` can work around this, albeit verbosely.

(Unfortunately this is an underspecified and unsound corner of Python static typing, since `Callable` doesn't explicitly encode either bound-method descriptor behavior or the lack of it, and it breaks substitutability in both directions. (Really descriptors in general are unavoidably unsound, since adding `__get__` returning any type other than `Self` should always be a Liskov violation, and `object` does not have `__get__`.))

We should either display this "function-like" marker somehow in the type display, so the displays here don't look identical, or make the flag entirely hidden (so the two types are treated as equivalent, even though they carry distinctive hints about how to bind as a descriptor). I don't love either option...

**Comment by alex** [comment]
Yeah, the circumstances that led to this definitely make sense.

(I went with simply switching to `Mapping`, since that was plenty fine for my use case.)

**Comment by pantheraleo-7** [comment]
strangely, passing an object directly doesn't result in an error:

```python
from collections.abc import Mapping


def f(d: dict[int, Mapping]): ...


f({1: {}})  # no error
a = {1: {}}  # assign to a variable
f(a)  # now it's an invariance error
```

**Comment by carljm** [comment]
@pantheraleo-7 Yes, this difference is expected, and you can find many similar examples, in ty and other Python type checkers. I recently explained this in the [second half of this comment](https://github.com/astral-sh/ty/issues/3672#issuecomment-4634790258).

**Comment by carljm** [comment]
I think the right fix here is probably to promote function-like callable types to regular callable types, whenever they occur in a generic specialization.

**Comment by carljm** [comment]
Hmm, this turns out to be a difficult semantic tradeoff decision. https://github.com/astral-sh/ruff/pull/27767 implements this widening, and successfully removes 14 diagnostics related to this confusing incompatibility between "container of function-like callables" and "container of regular callables". But it also creates 59 new diagnostics in cases where people put functions into a list/dict/set and then later pull them out and want to access e.g. `__name__` on them.

One option here is to say you should explicitly annotate your container as e.g. `list[Callable[...] & types.FunctionType]` if you want to restrict it to function-like callables. That's principled but not very useful to anyone writing code for multiple type checkers. It's also more appealing if we fix `FunctionType.__call__` to return `object` instead of `Any`, so calling that intersection still gives you the right return type, rather than an intersection with `Any`.

If we don't fix this issue, the workaround for people impacted by it is to explicitly annotate their container, with a spelled-out Callable type that does not imply function-like, e.g. the OP becomes

```py
from collections.abc import Callable

def f() -> list[str]:
    return []

def takes(names: dict[str, Callable[[], list[str]]]) -> None:
    pass

d: dict[str, Callable[[], list[str]]] = {"f": f}
takes(d)
```
which works fine.

I don't mind that workaround terribly much, but it feels very awkward to present the original error without the workaround in any way that seems reasonable. "Sorry we inferred this container as containing only function-like callables, if you want to put some other kind of callable in it, you need to add an annotation." Given that the typing spec and other type checkers make no such distinction, this feels unjustified.

@AlexWaygood what do you think we should do here?
